src/frequenz/dispatch/actor.py

```diff
@@ -0,0 +1,279 @@
+                    _logger.info("Updated dispatch: %s", dispatch)
+                    await self._updated_dispatch_sender.send(Updated(dispatch=dispatch))
+
+        except grpc.aio.AioRpcError as error:
+            _logger.error("Error fetching dispatches: %s", error)
+            self._dispatches = old_dispatches
+            return
+
+        for dispatch in old_dispatches.values():
+            _logger.info("Deleted dispatch: %s", dispatch)
+            await self._updated_dispatch_sender.send(Deleted(dispatch=dispatch))
+            if task := self._scheduled.pop(dispatch.id, None):
+                task.cancel()
+
+    def _update_dispatch_schedule(
+        self, dispatch: Dispatch, old_dispatch: Dispatch | None
+    ) -> None:
+        """Update the schedule for a dispatch.
+
+        Schedules, reschedules or cancels the dispatch based on the start_time
+        and active status.
+
+        For example:
+            * when the start_time changes, the dispatch is rescheduled
+            * when the dispatch is deactivated, the dispatch is cancelled
+
+        Args:
+            dispatch: The dispatch to update the schedule for.
+            old_dispatch: The old dispatch, if available.
+        """
+        if (
+            old_dispatch
+            and old_dispatch.active
+            and old_dispatch.start_time != dispatch.start_time
+        ):
+            if task := self._scheduled.pop(dispatch.id, None):
+                task.cancel()
+
+        if dispatch.active and dispatch.id not in self._scheduled:
+            self._scheduled[dispatch.id] = asyncio.create_task(
+                self._schedule_task(dispatch)
+            )
+
+    async def _schedule_task(self, dispatch: Dispatch) -> None:
+        """Wait for a dispatch to become ready.
+
+        Waits for the dispatches next run and then notifies that it is ready.
+
+        Args:
+            dispatch: The dispatch to schedule.
+        """
+
+        def next_run_info() -> tuple[datetime, datetime] | None:
+            now = datetime.now(tz=timezone.utc)
+            next_run = self.calculate_next_run(dispatch, now)
+
+            if next_run is None:
+                return None
+
+            return now, next_run
+
+        while pair := next_run_info():
+            now, next_time = pair
+
+            if next_time - now > _MAX_AHEAD_SCHEDULE:
+                await asyncio.sleep(_MAX_AHEAD_SCHEDULE.total_seconds())
+                continue
+
+            _logger.info("Dispatch %s scheduled for %s", dispatch.id, next_time)
+            await asyncio.sleep((next_time - now).total_seconds())
+
+            _logger.info("Dispatch ready: %s", dispatch)
+            await self._ready_dispatch_sender.send(dispatch)
+
+        _logger.info("Dispatch finished: %s", dispatch)
+        self._scheduled.pop(dispatch.id)
+
+    @classmethod
+    def calculate_next_run(cls, dispatch: Dispatch, _from: datetime) -> datetime | None:
```

Comment:
Agreed that is useful to have, I even wonder if we should move this method to the `Dispatch` class instead (so to the client), seems like something anyone using the API client directly probably want to have too, an easy way to know when a dispatch should be run next.

Reply:
* https://github.com/frequenz-floss/frequenz-client-dispatch-python/issues/28